packages/actor-http-cache/lib/ActorHttpCache.ts

```diff
@@ -0,0 +1,233 @@
+    await this.mediatorHttpInvalidate.mediate({ url: newRequest.url, context: action.context });
+    const revalidationHeaders = this.getRevalidationHeaders(newRequest, cacheResult.policy);
+    const response = await this.fetchDocument({
+      ...action,
+      init: {
+        ...action.init,
+        headers: revalidationHeaders,
+      },
+    });
+    const { policy, modified } = this.getRevalidatedPolicy(newRequest, response, cacheResult.policy);
+    const finalResponse = modified ? response : new Response(cacheResult.body, cacheResult.init);
+    return await this.teeAndCacheResponse(newRequest, finalResponse, policy, cacheStorage);
+  }
+
+  private async fetchDocument(action: IActionHttp): Promise<Response> {
+    const response = await this.mediatorHttp.mediate({
+      ...action,
+      context: action.context.set(KeysHttpCache.doNotCheckHttpCache, true),
+    });
+    return response;
+  }
+
+  private newCachePolicy(request: Request, response: Response): CachePolicy {
+    // Conversions to make headers compatible
+    const requestWithHashHeaders = ActorHttpCache.requestToRequestWithHashHeaders(request);
+    const responseWithHashHeaders = ActorHttpCache.responseToResponseWithHashHeaders(response);
+    return new CachePolicy(
+      requestWithHashHeaders,
+      responseWithHashHeaders,
+    );
+  }
+
+  private isPolicySatisfied(request: Request, policy: CachePolicy): boolean {
+    const requestWithHashHeaders = ActorHttpCache.requestToRequestWithHashHeaders(request);
+    return policy.satisfiesWithoutRevalidation(requestWithHashHeaders);
+  }
+
+  private getRevalidationHeaders(request: Request, policy: CachePolicy): Headers {
+    const requestWithHashHeaders = ActorHttpCache.requestToRequestWithHashHeaders(request);
+    const hashRevalidationHeaders = policy.revalidationHeaders(requestWithHashHeaders);
+    return new Headers(<Record<string, string>> hashRevalidationHeaders);
+  }
+
+  private getRevalidatedPolicy(request: Request, response: Response, policy: CachePolicy): RevalidationPolicy {
+    const requestWithHashHeaders = ActorHttpCache.requestToRequestWithHashHeaders(request);
+    const responseWithHashHeaders = ActorHttpCache.responseToResponseWithHashHeaders(response);
+    return policy.revalidatedPolicy(requestWithHashHeaders, responseWithHashHeaders);
+  }
+
+  private async teeAndCacheResponse(
+    request: Request,
+    response: Response,
+    cachePolicy: CachePolicy,
+    cacheStorage: IHttpCacheStorage,
+  ): Promise<Response> {
+    // Node-fetch does not support body.tee, while it is mandatory according to the fetch and readablestream api.
+    // If it doesn't exist, we monkey-patch it.
+    if (response.body && !response.body.tee) {
+      response.body.tee = (): [ReadableStream, ReadableStream] => {
+        // eslint-disable-next-line import/no-nodejs-modules
+        const Readable = require('stream').Readable;
+        return [
+          Readable.from(response.body),
```

Comment:
Not sure this way of cloning is fully safe.
Just learned about this package though, which looks like it meets all of our requirements: https://www.npmjs.com/package/cloneable-readable

Reply:
Unfortunately, the value of Body from the NodeJS fetch doesn't seem to be a stream. It's a `Buffer`. So, this library will not work. I'll look for other solutions. What are the requirements that need to be met?

Reply:
clonedStream

Reply:
I see. In order to make a `Response` object and maintain the stream that was passed in as the body, the `body instanceof Stream` must evaluate to `true`. This library is of type `Cloneable` which does not evaluate to stream. As discussed here (https://stackoverflow.com/questions/9898875/good-way-to-change-an-objects-prototype-to-change-results-of-instanceof), there is no way to modify the prototype of an object while maintaining its classes.

Reply:
> In order to make a Response object and maintain the stream that was passed in as the body, the body instanceof Stream must evaluate to true

Where does this cause an error exactly? Is it a typing error? Or a runtime error (in the Response constructor)?

Reply:
When you construct a `Request` object and pass in a stream, it will go here (https://github.com/node-fetch/node-fetch/blob/main/src/body.js#L53). However, if the object that you provide is not an `instanceof` stream, the code will hop here (https://github.com/node-fetch/node-fetch/blob/main/src/body.js#L59), and automatically turn the stream into a buffer. Because this is behavior built into the request object, it can't be changed. Perhaps we could build an entirely new request object that follows the same API. I'll try that and see if it works.

Reply:
Interestingly, the following works in a clean project, but doing the same thing in this project doesn't work. 

```
const stream = require("stream");
import "cross-fetch/polyfill";

const store = [];

async function main() {
  const response = await fetch("https://example.com");
  // @ts-ignore
  const stream1 = response.body.pipe(new stream.PassThrough());
  // @ts-ignore
  const stream2 = response.body.pipe(new stream.PassThrough());

  // @ts-ignore
  stream1.on("data", (chunk: any) => {
    console.log("chunk 1");
    console.log(chunk.toString());
  });
  // @ts-ignore
  stream1.on("end", () => {
    console.log("Stream 1 Ended");
  });

  setTimeout(() => {
    stream2.on("data", (chunk: any) => {
      console.log("chunk 2");
      console.log(chunk.toString());
    });
    stream2.on("end", () => {
      console.log("Stream 2 Ended");
    });
  }, 5000);
}

main();
```

Reply:
@rubensworks I've done some experimenting with stream cloning techniques. In isolation, this one seems to work (https://github.com/jaxoncreed/clone-stream-demo). You can clone this repo and check it yourself.

But, implementing this solution inside communica causes problems. I've created a demo of that here (https://github.com/jaxoncreed/comunica/pull/1/files). The `end` event is never called if two streams are created. (You can see that it is successful by removing the second clone line here https://github.com/jaxoncreed/comunica/pull/1/files#diff-032bb79de64c61798f786bec348f8bbabe74a11da61cafbde03f44aeb797305eR92)

I can't figure out why this solution works in the first codebase, but not the second. Do you have any ideas?

Reply:
Most tests use a minimal body mock: https://github.com/jaxoncreed/comunica/blob/9eddbeae4488a5e0e349d4ecb7161cc7935a0535/packages/actor-http-fetch/test/ActorHttpFetch-test.ts#L17

For tests that don't override the mocked body, I guess this could explain hanging streams?

Reply:
@rubensworks I'll need to update those tests in the future, so I agree that the unit tests for the project are a problem, but the problem persists with integrations tests as well. For example, I'm running the following to test the app: `jest engines/query-sparql/test/QuerySparql-test.ts -t "for the single source context entry" --coverage false` and it still hangs.

Reply:
I vaguely remember hanging Node problems when the body was left unconsumed.
Perhaps the problem is that the cloned and cached body is never consumed, which makes Node hang?
You can probably easily test this by removing the data and end listeners on the second stream of your clone-stream-demo.

If this is the problem, then we can't really force stream consumption after end I think, as this would defeat the purpose of the cache. But perhaps there is a way to make that stream not hang.

Reply:
Ah, indeed, that is the case. Removing the stream listeners for the second stream will cause the demo to hang for requests that have more than one chunk.

But, there's some more bad news. The problem goes beyond the fact that `stream2` keeps Node running. It also keeps the the subsequent chunks from triggering events on `stream1`.

For example given the following code and a stream with 3 chunks...

```typescript
  const stream1 = response.body.pipe(new stream.PassThrough());
  const stream2 = response.body.pipe(new stream.PassThrough());

  let stream1ChunkCount = 0;
  stream1.on("data", (chunk: any) => console.log(`stream1 chunk ${++stream1ChunkCount}`));
  stream1.on("end", () => console.log("Stream 1 Ended"));

  setTimeout(() => {
    let stream2ChunkCount = 0;
    stream2.on("data", (chunk: any) => console.log(`stream2 chunk ${++stream2ChunkCount}`));
    stream2.on("end", () => console.log("Stream 2 Ended"));
  }, 5000);
```

This is logged
```
stream1 chunk 1
<-- Waits for 5 seconds -->
stream1 chunk 2
stream2 chunk 1
stream2 chunk 2
stream1 chunk 3
stream2 chunk 3
Stream 1 Ended
Stream 2 Ended
```

So, the streams, while cloned, are still intertwined.

Reply:
> But, there's some more bad news. The problem goes beyond the fact that stream2 keeps Node running. It also keeps the the subsequent chunks from triggering events on stream1.

I see that you're snippet is using the `PassThrough`-approach for cloning, so this behaviour makes sense.

Wasn't it possible to make it work using `cloneable-readable`? Because I suspect that one should solve all of the problems we have. We just may have to create a custom `Stream` wrapper for it ourselves.

Reply:
No the `cloneable-readable` causes the same result. The stream will not trigger events until both the clone and the original have registered their listeners:

For example, the following code

```typescript
const cloneableBody = cloneable(response.body);
  const stream1 = cloneableBody.clone();
  const stream2 = cloneableBody;

  let stream1ChunkCount = 0;

  stream1.on("data", (chunk: any) => {
    console.log(`stream1 chunk ${++stream1ChunkCount}`);
  });
  stream1.on("end", () => {
    console.log("Stream 1 Ended");
  });

  setTimeout(() => {
    console.log('Timeout')
    let stream2ChunkCount = 0;
    stream2.on("data", (chunk: any) => {
      console.log(`stream2 chunk ${++stream2ChunkCount}`);
    });
    stream2.on("end", () => {
      console.log("Stream 2 Ended");
    });
  }, 5000);
```
prints
```
Timeout Complete
stream1 chunk 1
stream2 chunk 1
stream1 chunk 2
stream2 chunk 2
stream1 chunk 3
stream2 chunk 3
Stream 2 Ended
Stream 1 Ended
```

Stream 1 doesn't start sending chunks until after the timeout is complete

Reply:
@rubensworks what do you think we should do? It doesn't look possible to keep a stream suspended while using another stream.

Reply:
Hmm, that's too bad. I thought `cloneable-readable` would enable this.

I would suggest to solve the problem by fully consuming the stream in `HttpCacheStorageLru`, and storing it directly (e.g. via a `Buffer`). So I would keep the `cloneable-readable` logic, as we'll need to duplicate the stream in any case. But instead of storing the stream, we store the data inside the stream. And when the data is requested from the cache, we replay the stream.

Some things to keep in mind for this:
- We should make sure that the replayed streams are lazy, and don't start flowing before data listeners are attached. This is to avoid expensive streams to start if only things like headers or so are read from the cache.
- We should add a (configurable?) limit on the number of bytes we store from streams. And when the body stream exceeds that threshold, we should just not cache this request. This is to avoid memory issues with very large or infinite files.
- We should be able to handle cases where a cached body is requested while the cached body is still in progress of being read. We can do this by storing the body in 2 phases. 1) While the body that we are caching is being read to store in the `HttpCacheStorageLru`, we temporarily store a reference to this stream, so that we can make another clone of it when it is being read from cache. 2) Once the body stream that we are caching has ended, we stop storing this stream's reference, and instead store the bytes directly, so that we can replay it as a stream once retrieved from cache.

Does this make sense?

Reply:
Okay. I'll figure it out.

Reply:
All right, I whipped up a demo that I think satisfies all the requirements. @rubensworks  would you mind looking it over to confirm that this is what you want me to do? If that looks good, I'll get to work on refactoring and writing tests:

https://github.com/jaxoncreed/clone-stream-demo/blob/buffer-solution/lib/index.ts

This code logs the following:
```
Request made
stream1 chunk 1
Short Term Cached value is needed
shortTermCachedStream chunk 1
stream1 chunk 2
shortTermCachedStream chunk 2
stream1 chunk 3
shortTermCachedStream chunk 3
Stream 1 Ended
shortTermCachedStream Ended
Request is too big to store
Timeout complete. Long Term Cached value is needed
longTermCachedStream chunk 1
longTermCachedStream Ended
All stream values are the same: true
```

Reply:
@jaxoncreed The approach in your demo looks very good to me!

Some thoughts:
- I guess we can use `cloneable-readable` instead of passthrough-based cloning.
- The short-term caching approach is probably very sensitive to bugs, so we'll need some good unit tests to check all edge-cases.

Reply:
Sounds good. A few quick questions:
* What is wrong with using the "stream" pass-through solution? Is it because of concerns over NodeJS libraries being available in browser, in which case I suggest using the [readable stream library](https://www.npmjs.com/package/readable-stream).
* Here's a list of the edge cases that I think need to be tested for the short term cache. Can you think of anymore?:
   * Cache needed when the request has been sent, but even the headers haven't been returned yet (synchronous - the request for the cache was made in the same closure as the first request)
   * Cache needed when the request has been sent, but even the headers haven't been returned yet (asynchronous - the request for the cache was made in a different closure separated by some kind of 0second await)
   * Cache needed when the request has come in, and there is no body (synchronous)
   * Cache needed when the request has come in, and there is no body (asynchronous)
   * Cache is needed when the first part of a single chunk body has come in (synchronous)
   * Cache is needed when the first part of a single chunk body has come in (asynchronous)
   * Cache is needed when the first part of a multi-chunk body has come in (synchronous)
   * Cache is needed when the first part of a multi-chunk body has come in (asynchronous)
   * Cache is needed when the middle part of a multi-chunk body has come in (synchronous)
   * Cache is needed when the middle part of a multi-chunk body has come in (asynchronous)
   * Cache is needed when the final part of a multi-chunk body has come in (synchronous)
   * Cache is needed when the middle part of a multi-chunk body has come in (asynchronous)